packages/bpk-component-image/src/withLazyLoading.js

```diff
@@ -15,64 +15,92 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+/* @flow */
 
-import React from 'react';
+import React, { type ComponentType } from 'react';
 import PropTypes from 'prop-types';
 import { wrapDisplayName } from 'bpk-react-utils';
 
-export default function withLazyLoading(Component, document) {
-  class WithLazyLoading extends React.Component {
-    constructor() {
+type WithLazyLoadingProps = {
+  className: ?string,
+  style: ?{},
+};
+
+type WithLazyLoadingState = {
+  inView: boolean,
+};
+
+export default function withLazyLoading(
+  Component: ComponentType<any>,
+  documentRef: Document,
+): ComponentType<any> {
+  class WithLazyLoading extends React.Component<
+    WithLazyLoadingProps,
+    WithLazyLoadingState,
+  > {
+    checkInView: Function;
+    element: HTMLElement;
+    isInViewPort: Function;
+    removeEventListeners: Function;
+    setInView: Function;
+    supportsPassiveEvents: Function;
+    state: WithLazyLoadingState;
+    placeholderReference: string;
+
+    static defaultProps: {};
+
+    constructor(): void {
       super();
 
       this.setInView = this.setInView.bind(this);
       this.removeEventListeners = this.removeEventListeners.bind(this);
       this.checkInView = this.checkInView.bind(this);
       this.isInViewPort = this.isInViewPort.bind(this);
       this.supportsPassiveEvents = this.supportsPassiveEvents.bind(this);
-
       this.state = {
         inView: false,
       };
     }
 
-    componentDidMount() {
-      const passiveArgs = this.supportsPassiveEvents() ? { passive: true } : {};
-      document.addEventListener('scroll', this.checkInView, {
+    componentDidMount(): void {
+      documentRef.addEventListener('scroll', this.checkInView, {
         capture: true,
-        ...passiveArgs,
+        ...this.getPassiveArgs(),
       });
-      document.addEventListener('resize', this.checkInView);
-      document.addEventListener('orientationchange', this.checkInView);
-      document.addEventListener('fullscreenchange', this.checkInView);
+      documentRef.addEventListener('resize', this.checkInView);
+      documentRef.addEventListener('orientationchange', this.checkInView);
+      documentRef.addEventListener('fullscreenchange', this.checkInView);
       // call checkInView immediately incase the
       // component is already in view prior to scrolling
       this.checkInView();
     }
 
-    componentWillUnmount() {
+    componentWillUnmount(): void {
       this.removeEventListeners();
     }
 
-    setInView() {
-      this.setState(() => ({
+    setInView(): void {
```

Comment:
I'm surprised this and the below line required typing. Seems like in both cases Flow would infer these. Generally it looks like you are having to add a lot more types than I've had to in the code I've flow typed

Reply:
I don't have to add that one technically, I went too deep in the rabbit hole